controllers/admin/AdminCartRulesController.php

```diff
@@ -60,6 +60,97 @@ public function __construct()
         ];
     }
 
+    /**
+     * Function used to render the list to display for this controller.
+     * Add joins to show only cart_rules to which the employee has access, restrict edit and delete action
+     *
+     * @return string|false
+     *
+     * @throws PrestaShopException
+     */
+    public function renderList()
+    {
+        if (!($this->fields_list && is_array($this->fields_list))) {
+            return false;
+        }
+
+        $context = Context::getContext();
+        $all_shops = Shop::getCompleteListOfShopsID();
+        if ($context->employee->isSuperAdmin()) {
+            $employee_shops = $all_shops;
+        } else {
+            $employee_shops = $context->employee->getAssociatedShops();
+        }
+        $context_shops = Shop::getContextListShopID();
+        $employee_has_full_access = empty(array_diff($all_shops, $employee_shops));
+        $this->_select = ($employee_has_full_access ? '0' : '(cse.id_cart_rule is not null or cs.id_cart_rule is null)') . ' as restricted';
+        $this->_join = 'left join `' . _DB_PREFIX_ . 'cart_rule_shop` cs on cs.id_cart_rule = a.id_cart_rule ' . // if there is none, the cart_rule is in all shops
+            'left join `' . _DB_PREFIX_ . 'cart_rule_shop` csc on csc.id_cart_rule = a.id_cart_rule and csc.id_shop in (' . implode(', ', $context_shops) . ') ' . // shops in current context where employee has access
+            'left join `' . _DB_PREFIX_ . 'cart_rule_shop` cse on cse.id_cart_rule = a.id_cart_rule and cse.id_shop not in (' . implode(', ', $employee_shops) . ') '; // if a cart_rule belongs to a shop where the employee has no access then restrict
+        $this->_where = 'and (case when cs.id_cart_rule is null then 1 else (csc.id_cart_rule is not null) end) = 1';
+        $this->_group = 'group by a.id_cart_rule';
+
+        $this->getList($this->context->language->id);
+        $restricted = [];
+        foreach ($this->_list as $row) {
+            if ($row['restricted']) {
+                $restricted[] = $row['id_cart_rule'];
+            }
+        }
+
+        // If list has 'active' field, we automatically create bulk action
+        if (isset($this->fields_list) && is_array($this->fields_list) && array_key_exists('active', $this->fields_list)
```

Comment:
The is_array check has already been done above :thinking:

Reply:
Lines 93 - 136 and 140 - 152 are copied as is from `AdminController.php`. I prefer you discuss my additions, rather than preexisting code. 
I'm reluctant to copying code, but here I didn't see any way, I could override this function. It's because of the way the `Helper` object is instantiated.

Reply:
Since it's an addition, we are for applying the boy scout rules, make it better than it was ^^

Reply:
_Boy scout rules_ sounds good...

Reply:
@PierreRambaud see #28171 for applying your recommendation to `AdminController.php`

Reply:
Thank you @lmeyer1 :pray: